### PR TITLE
GROK-13240: ML: Parallel Distance matrix calculation

### PR DESCRIPTION
**Pull request by drizhina**
Things done in this PR:

- Introduce DistanceMatrixService, which allows to calculate distance matrix on multiple workers. speedups up to 10x depending on CPU. 
- Use Condensed distance matrix(Float32Array) for dimensionality reduction methods instead of full nXn matrix.
- Fix tests
Results:
Dendrogram: clustering calculation speedup up to 10X on macromolecules.
Bio: Sequence space and activity cliffs calculation speedup ~5X. sim/div search speedup ~5-6X
Chem: unchanged. for now can not paralelize matrix calculation for BitArray, as webworkers can't be passed custom classes(coming in future)

### CODE REVIEW

File: libraries/ml/src/distance-matrix/distance-matrix-service.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+          let lmin = 0;
+          let lmax = Number.MIN_VALUE;
+          for (let i = 0; i < this._workerCount; i++) {
+            const start = Math.floor(i * chunkSize);
+            const end = (i === this._workerCount - 1) ? totalLength : Math.floor((i + 1) * chunkSize);
+            const startRow = endRow;
+            const startCol = endCol;
+            if (i !== this._workerCount - 1) {
+              // These formulas map the linear index to the upper triangular matrix indices
+              endRow = len - 2 - Math.floor(Math.sqrt(-8 * end + 4 * len * (len - 1) - 7) / 2 - 0.5);
```
Comment (LaykDimon) [inline libraries/ml/src/distance-matrix/distance-matrix-service.ts:37]: guess you better put the numbers into constants, as the time will pass and you will forget what these numbers mean
and if you put it into named constants, you'll (and not only you) will recognize it much more better

File: libraries/ml/src/distance-matrix/distance-matrix.ts [hunk 2]
```diff
@@ -58,8 +58,11 @@ export class DistanceMatrix {
     const size: number = list.length;
     const res = new DistanceMatrix(undefined, size);
     for (let i = 0; i < size; i++) {
-      for (let j = i + 1; j < size; j++)
-        res.set(i, j, method(list[i], list[j]));
+      for (let j = i + 1; j < size; j++) {
+        // if any of the values is null, set distance to 1
+        list[i] !== null && list[j] !== null ?
+          res.set(i, j, method(list[i], list[j])) : res.set(i, j, 1);
```
Comment (LaykDimon) [inline libraries/ml/src/distance-matrix/distance-matrix.ts:64]: you can do it even better
make the condition inside your set() method - `res.set(i, j, ..... ? method(list[I], list[j]) : 1)`

File: libraries/ml/src/reduce-dimensionality.ts [hunk 3]
```diff
@@ -351,20 +382,20 @@ export class DimensionalityReducer {
...
    * @param {boolean} transpose Whether to transform coordinates to have columns-first orientation.
+   * @param {boolean} parallelDistanceWorkers Whether to use parallel distance computation.
    * @throws {Error} If the embedding method was not found.
    * @return {any} Cartesian coordinate of this embedding and distance matrix where applicable.
    * @memberof DimensionalityReducer
    */
-  public transform(transpose: boolean = false): { [key: string]: Matrix } {
-    if (this.reducer == undefined) {
+  public async transform(transpose: boolean = false, parallelDistanceWorkers?: boolean): Promise<IReduceDimensionalityResult> {
+    if (this.reducer == undefined)
```
Comment (LaykDimon) [inline libraries/ml/src/reduce-dimensionality.ts:391]: strict comparison

File: libraries/ml/src/spe.ts [hunk 4]
```diff
@@ -84,9 +89,9 @@ export class SPEBase {
 
     let lambda2 = this.lambda2;
 
-    if (this.steps == 0) {
-      this.steps = vectors.length-1;
-    }
+    if (this.steps == 0)
```
Comment (LaykDimon) [inline libraries/ml/src/spe.ts:92]: strict comparison

File: libraries/ml/src/spe.ts [hunk 5]
```diff
@@ -213,19 +216,18 @@ export class OriginalSPE extends SPEBase {
 
     this.initDistance(vectors);
 
-    if (this.maxDistanceSteps == null) {
+    if (this.maxDistanceSteps == null)
```
Comment (LaykDimon) [inline libraries/ml/src/spe.ts:219]: same

File: libraries/ml/src/viewers/activity-cliffs.ts [hunk 6]
```diff
@@ -90,13 +90,13 @@ export async function getActivityCliffs(df: DG.DataFrame, seqCol: DG.Column, enc
...
   linesGridFunc?: (df: DG.DataFrame, pairColNames: string[]) => DG.Grid,
   seqSpaceOptions?: any, cliffsDockRatio?: number) : Promise<DG.Viewer> {
-
   activityCliffsIdx++;
   const similarityLimit = similarity / 100;
   const dimensionalityReduceCol = encodedCol ?? seqCol;
   let zoom = false;
   let ignoreSelectionChange = false;
   const cashedLinesData: any = {};
+  // eslint-disable-next-line prefer-const
```
Comment (LaykDimon) [inline libraries/ml/src/viewers/activity-cliffs.ts:99]: why is it done?

File: libraries/ml/src/distance-matrix/distance-matrix-service.ts (hunk 1 above)
Comment (drizhina) [reply libraries/ml/src/distance-matrix/distance-matrix-service.ts:37]: I don't think that these numbers have any actual meaning. they are just some constants that pop out when you start working out how to map a linear indexed compressed distance matrix to a square one.

File: libraries/ml/src/reduce-dimensionality.ts [hunk 7]
```diff
@@ -80,14 +85,15 @@ abstract class Reducer {
...
+  abstract transform(parallelDistanceWorkers?: boolean): Promise<IReduceDimensionalityResult>;
 }
 
 /** t-SNE dimensionality reduction. */
 class TSNEReducer extends Reducer {
   protected reducer: TSNE;
   protected iterations: number;
-  protected distanceFn: DistanceMetric;
+  protected distanceFname: KnownMetrics;
+  protected distanceFn: (a: any, b: any) => number;
```
[the remark this answers is not in the document]
Comment (drizhina) [reply libraries/ml/src/reduce-dimensionality.ts:96]: yes. distance function will take two arbitrary things and compute a distance. as far as reducer is concerned, they can be anything, strings, arrays, bitarrays, numbers...

File: libraries/ml/src/distance-matrix/distance-matrix-service.ts (hunk 1 above)
Comment (LaykDimon) [reply libraries/ml/src/distance-matrix/distance-matrix-service.ts:37]: if we can't find their real meaning it's ok so

File: libraries/ml/src/reduce-dimensionality.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (skalkin) [reply libraries/ml/src/reduce-dimensionality.ts:96]: How about 

```
protected distanceFn<T>: (a: T, b: T) => number
```
?

File: libraries/ml/src/spe.ts (hunk 4 above)
Comment (skalkin) [reply libraries/ml/src/spe.ts]: What's wrong with it though? We know that this.steps is a number, and we'd get a warning if we try to compare incompatible types using "=="

**Review by skalkin — APPROVED** [review]

File: libraries/ml/src/spe.ts (hunk 4 above)
Comment (LaykDimon) [reply libraries/ml/src/spe.ts]: why do we even need to use '==' in JS?
strict comparison is safer and faster (doesn't use type casting)

**Pull request merged**